source/extensions/filters/http/admission_control/admission_control.h

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/common/time.h"
+#include "envoy/extensions/filters/http/admission_control/v3alpha/admission_control.pb.h"
+#include "envoy/http/codes.h"
+#include "envoy/http/filter.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/common/cleanup.h"
+#include "common/grpc/common.h"
+#include "common/grpc/status.h"
+#include "common/http/codes.h"
+#include "common/runtime/runtime_protos.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+/**
+ * All stats for the admission control filter.
+ */
+#define ALL_ADMISSION_CONTROL_STATS(COUNTER) COUNTER(rq_rejected)
+
+/**
+ * Wrapper struct for admission control filter stats. @see stats_macros.h
+ */
+struct AdmissionControlStats {
+  ALL_ADMISSION_CONTROL_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/**
+ * Thread-local object to track request counts and successes over a rolling time window. Request
+ * data for the time window is kept recent via a circular buffer that phases out old request/success
+ * counts when recording new samples.
+ *
+ * The look-back window for request samples is accurate up to a hard-coded 1-second granularity.
+ * TODO (tonya11en): Allow the granularity to be configurable.
+ */
+class ThreadLocalController {
```

Comment:
Yeah given the size of this PR if possible maybe split the controller out into its own PR? If it's a big deal I can do in one review but it would be nice to not have so much code to look at.